### PR TITLE
Added EventTimeline graph to memory profiler.

### PR DESCRIPTION
**Pull request by terrylucas**
Added the Event timeline to the memory profile.  When "Snapshot" or "Reset" button is clicked the "Event Timeline" subplot is displayed and time of the event is record .  Reset's are semi-attached to the last snapshot (the accumulators are reset to zero).

To do allow charts to be re-sized and retrieve history snapshot/accumulator values - click on the snapshot marker, in the event chart, to fetch the old values at that timestamp.

The event timeline can be added to any chart and the contents are also dynamic (vertical grows).

Supports both light and dark mode.

image

Also, added proper font-family to match fonts on the memory page.

### CODE REVIEW

File: packages/devtools/lib/src/memory/memory_plotly.dart [hunk 1]
```diff
@@ -105,8 +216,8 @@ class MemoryPlotly {
 
     if (marker == null) {
       return Data(
-        x: [],
-        y: [],
+        x: [Null], // Trace legend entry appears w/o data.
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory_plotly.dart]: Nit: Trace legend entry appears w/o data --> Null is needed so the trace legend entry appears w/o data.

File: packages/devtools/lib/src/memory/memory_plotly.dart [hunk 2]
```diff
@@ -254,4 +379,140 @@ class MemoryPlotly {
...
+
+    final List data = getProperty(_memoryChart.element, 'data');
+    final Data capacityTrace = data[MEMORY_CAPACITY_TRACE];
+    final int timestamp = capacityTrace.x[capacityTrace.x.length - 1];
+
+    eventTimeline.plotReset(timestamp);
+  }
+}
+
+class EventTimeline {
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory_plotly.dart]: add a comment or make the name more clearly memory related?

File: packages/devtools/lib/src/memory/memory_protocol.dart [hunk 3]
```diff
@@ -60,6 +60,7 @@ class MemoryTracker {
     }
 
     final VM vm = await service.getVM();
+    // TODO(terry): Need to handle a possible Sentinel being returned.
     final List<Isolate> isolates =
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory_protocol.dart]: are you getting sentinels for the isolates? That is a bit strange unless the isolate has shut down or something.

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart [hunk 4]
```diff
@@ -40,7 +40,7 @@ class FramesBarChart extends CoreElement with SetStateMixin {
     });
   }
 
-  static const int chartHeight = 160;
+  static const int chartHeight = 230;
```
Comment (jacob314) [inline packages/devtools/lib/src/timeline/frames_bar_chart.dart]: hmm. this is a significant increase. @kenzieschmoll does making the height be this high work for you or do we need to just have the views use different heights?

**Comment by terrylucas** [comment]
Since I don't create the event timeline until snapshot is pressed I could
delay sizing the memory chart until then.

**Comment by jacob314** [comment]
Having the timeline only expand when needed sounds like a great compromise.

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart (hunk 4 above)
Comment (kenzieschmoll) [reply packages/devtools/lib/src/timeline/frames_bar_chart.dart]: I'm not crazy about increasing the height of the FPS chart. This further reduces the space we have for the flame chart (which may become significantly taller soon if we can add cpu sampling to the flame chart). The FPS chart has the least granularity on the page, so it doesn't need a ton of space to get its message across. The flame chart (and maybe details section if we keep it), however, need to show a lot of data at once, so I think we should prioritize the real estate on the page for this (these) sections.

 Can we pass chart height in as a parameter? That way we can have different sizes for each page.

**Comment by kenzieschmoll** [comment]
Why do the FPS chart and the Memory chart need to be the same size? I don't think we should sacrifice usability for consistency, and I'm concerned about taking up too much space on the timeline page with the FPS chart. The memory chart is more complex than the FPS chart, so it makes sense that it would need more space.

**Comment by devoncarew** [comment]
I had made them the same height in the past in order for the different pages to better have simialr designs and layouts, and for there being less layout differences as you switch between pages. If there's a clear need for the graphs to be different heights, I'd just remove the common height constant and have each chart have a different height.

**Comment by kenzieschmoll** [comment]
Terry, if your workaround works smoothly with the rest of the memory page, then that sounds like a good compromise to me as well 👍

File: packages/devtools/lib/src/memory/memory_plotly.dart (hunk 1 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory_plotly.dart]: Done.

File: packages/devtools/lib/src/memory/memory_plotly.dart (hunk 2 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory_plotly.dart]: Added comment:

/// Create an Event Timeline subplot, notice it is associated with y2.  This
/// requires that the layout these traces exist in theyaxis2 area.  For an
/// example, look at MemoryPloty's getMemoryLayout method it creates a yaxis2
/// positioned above yaxis.

File: packages/devtools/lib/src/memory/memory_protocol.dart (hunk 3 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory_protocol.dart]: I did see it once - I think?  Look at trying to repro or add temporary logic to catch it if it happens again.

File: packages/devtools/lib/src/framework/framework.dart [hunk 5]
```diff
@@ -121,6 +124,17 @@ class Framework {
       mainElement.add(screenContent);
 
       _screenContents[current] = screenContent;
+
```
Comment (jacob314) [inline packages/devtools/lib/src/framework/framework.dart]: Add a comment that this is stopping event listeners within the screen from getting the resize event not stop event listeners higher up in the tree. I misunderstood that on first read. Give Plotly as a concrete example of why this is needed.

**Pull request merged**